### ISSUE TITLE
[Awaiting Checklist Completion] [$1000] After going back from a thread, clicking on New Message button doesn't scroll to the new message at the bottom

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21603]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open any LONG chat
2. Receive a NEW message (or just mark the latest message as UNREAD)
3. Scroll up the chat list >> Go into any sub-thread
4. Click Back button on Mobile or Browser to go back to the main chat
5. Click on New message button

## Expected Result:
 It should scroll to the latest/new message

## Actual Result:
Clicking on New Message button doesn't scroll to new message at the bottom

## Workaround:
Uknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.32.5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/a9376ae5-86ea-40de-8cd6-45ea7798a51a

https://github.com/Expensify/App/assets/93399543/94241b56-9612-40bf-ab24-516dcaba485b

https://github.com/Expensify/App/assets/93399543/8d775b96-b077-4ac2-9504-6a30ad2af791

**Expensify/Expensify Issue URL:**
**Issue reported by:** @tranvantoan-qn
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687451519479619

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0172cbe3d50bd00465</li>
        <li>Upwork Job ID: 1673464192182222848</li>
        <li>Last Price Increase: 2023-07-03</li>
    </ul>
</details>

**Comment by twisterdotcom** [comment]
I recreated this on desktop too. This is an annoying bug for sure. We must be somehow marking the room slightly as read when entering a thread?

**Comment by sabhyap** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

New Message popup button does not work if you scroll up receive a new message and you go inside a sub-thread and come back in the main thread. 

### What is the root cause of that problem?

The root cause of problem is caused because of the following line of code:
https://github.com/Expensify/App/blob/5f5dd023a8102ef47ac4b58e4bfe79b995cb0330/src/pages/home/report/FloatingMessageCounter/index.js#L25C4-L25C23

Basically, When we click the New Message button the page is re-rendered. This works fine until we click the browser's back button because React still thinks we are on the old page and re-renders that page instead of the currently opened page.

### What changes do you think we should make in order to solve the problem?

We can fix this by adding code to first navigate to current page and then re-render the page. This way it will always work with the current page.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by twisterdotcom** [comment]
I am OOO! Adding another BZ for the week.

**Comment by sabhyap** [comment]
Hi @thesahindia , I have fixed it on my local instance by changing the following code:
https://github.com/Expensify/App/blob/5f5dd023a8102ef47ac4b58e4bfe79b995cb0330/src/pages/home/report/FloatingMessageCounter/index.js#L25C4-L25C23

to 

`onClick: () => {this.props.isActive = false},`

-----

This was happening due to the same issue I told ealier. (The React state on Chromium / Chrome behaves weird when history is set back. We can fix this by updating the isActive state to False).

**Comment by lazar-stamenkovic** [comment]
### Please re-state the problem that we are trying to solve in this issue.
After going back from a thread, clicking on New Message button doesn't scroll to the new message at the bottom
What is the root cause of that problem?

### **The root cause of problem is caused because of the following line of code:**
In ReportActionsView,  the Report.subscribeToNewActionEvent function isn't subscribed after backing from any other screen.

i.e: 
When the user add any chat, notifyNewAction function is called for the current chat channel of report id.
![Screenshot 2023-07-04 at 08 48 56](https://github.com/Expensify/App/assets/114150992/a59a2d92-a229-4b00-948b-45bc362e25bd)

 When the this newActionSubscriber is subscribed In ReportActionsView, the scrolling works.
https://github.com/Expensify/App/blob/5f5dd023a8102ef47ac4b58e4bfe79b995cb0330/src/pages/home/report/ReportActionsView.js#L102

If the user go to sub-thread, report id of newActionSubscriber in Report.js is updated to report id of sub-thread. 
so notifyNewAction function works for sub-thread.

When backing to main screen,  report id of newActionSubscriber has to be updated to report id of main screen. 
but it doesn't work. so the scrolling doesn't work.



### What changes do you think we should make in order to solve the problem?
We can fix this issue followings.
Report.subscribeToNewActionEvent function move from componentDidMount to Report.subscribeToNewActionEvent.
Report.subscribeToNewActionEvent in Report.subscribeToNewActionEvent is called on following condition.
```
        if (Report.newActionSubscriber.report !== this.props.report.reportID) {
                          this.unsubscribeFromNewActionEvent = Report.subscribeToNewActionEvent(this.props.report.reportID, (isFromCurrentUser, newActionID) => {
                  .....
        }
```



What alternative solutions did you explore? (Optional)

**Comment by lazar-stamenkovic** [comment]
@tjferriss, @twisterdotcom, @thesahindia  
@sabhyap just updated his proposal after checking my proposal. lol. bad proposal flow.

**Comment by twisterdotcom** [comment]
Hi @sabhyap I do see what @lazar-stamenkovic is saying here. This is your original proposal:

<img width="967" alt="image" src="https://github.com/Expensify/App/assets/9133401/0b73c2e5-5c59-4dce-9f3b-58e815e8e60e">

Later, @lazar-stamenkovic updated their original proposal:
<img width="974" alt="image" src="https://github.com/Expensify/App/assets/9133401/fe67547d-7b05-4d7b-8e55-0821dbd01009">

And you subsequently updated yours to this:
<img width="1001" alt="image" src="https://github.com/Expensify/App/assets/9133401/124f5a06-7503-4be7-9d49-b4e3090a5d15">

Please **do not do this** and definitely not without an explanation. That is a clear violation of our Contributor Guidelines. There are plentiful times where solutions will use the thinking of other contributors, but we should always attribute that, even if a best solution ends up pulling from many.

@thesahindia I think we should probably go ahead with @lazar-stamenkovic if you agree with the solution.

**Comment by lazar-stamenkovic** [comment]
@twisterdotcom thanks for considering my application.
@thesahindia would you please take a look my proposal? 
https://github.com/Expensify/App/issues/21603#issuecomment-1620319845
I've also applied for the job associated with this ticket on Upwork.

**Comment by twisterdotcom** [comment]
@thesahindia we only have one proposal. Is it good enough or can we double it?

**Comment by lazar-stamenkovic** [comment]
@thesahindia @twisterdotcom Could you kindly assign me this ticket? hope to make my first pull request on this project and contribute to its development.

**Comment by deetergp** [comment]
You should be ready to proceed, @lazar-stamenkovic 👍

**Comment by lazar-stamenkovic** [comment]
Thanks @deetergp, i accepted the offer on the upwork. Will create a PR shortly.

**Comment by tranvantoan-qn** [comment]
@deetergp can you send me the offer for the reporter role? My Upwork account is new so it is set to private - so the GH workflow might not work for me

**Comment by tranvantoan-qn** [comment]
Contributor details
Your Expensify account email: tranvantoan.qn@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~012f07ead7068a224c

**Comment by lazar-stamenkovic** [comment]
@deetergp @twisterdotcom @thesahindia  After I submitted the proposal, someone migrated ReportActionsView.js from a class to a function. So I couldn't use my solution. I solved this bug perfectly with the following steps:

•  Use a new action subscriber list instead of a single new action subscriber object in Report.js
•  Append a new action subscriber object to the new action subscriber list when ReportActionsView is rendered.
•  Remove the new action subscriber object from the new action subscriber list when ReportActionsView is unmounted

I created a PR for this. 
https://github.com/Expensify/App/pull/22645

Please take a look.

**Comment by lazar-stamenkovic** [comment]
@thesahindia would you please review my PR for this ticket?   I created the PR yesterday. hope to complete the PR in 3 business days. Thanks.

**Comment by lazar-stamenkovic** [comment]
@allroundexperts I've already accepted a job offer for this ticket on the Upwork and created a PR #22645. 
so hope to fix this bug on this ticket.  will put a proposal again.

**Comment by lazar-stamenkovic** [comment]
### Please re-state the problem that we are trying to solve in this issue.
After going back from a thread, it doesn't scroll to the new message at the bottom when sending a new message.

### What is the root cause of that problem?
this scrolling works when subscribeToNewActionEvent is called in ReportActionView.js
https://github.com/Expensify/App/blob/138a41c745761898867e14fd199940f34875d3ac/src/pages/home/report/ReportActionsView.js#L133

The app uses a newActionSubscriber property in Reports.js to subscribe new message sending action.
https://github.com/Expensify/App/blob/138a41c745761898867e14fd199940f34875d3ac/src/libs/actions/Report.js#L180

If the user go to sub-thread, report id of newActionSubscriber in Report.js is updated to report id of sub-thread. 
so notifyNewAction function works for sub-thread.

newActionSubscriber value is cleared when the app backs subthread.
https://github.com/Expensify/App/blob/138a41c745761898867e14fd199940f34875d3ac/src/pages/home/report/ReportActionsView.js#L155C1-L156C1
https://github.com/Expensify/App/blob/138a41c745761898867e14fd199940f34875d3ac/src/libs/actions/Report.js#L192

When backing to main screen,  report id of newActionSubscriber has to be updated to report id of main screen. 
but it doesn't work. so the scrolling doesn't work for main screen.


### What changes do you think we should make in order to solve the problem?
• Use a new action subscriber list instead of a single new action subscriber object in Report.js
• Append a new action subscriber object with new report action id to the new action subscriber list whenever a new report screen is rendered.
• Remove the new action subscriber object from the new action subscriber list when ReportActionsView is unmounted





What alternative solutions did you explore? (Optional)

**Comment by lazar-stamenkovic** [comment]
> I get that @lazar-stamenkovic but just want to confirm with the team so that we have the correct history of solved bugs.

Thanks @allroundexperts

**Comment by twisterdotcom** [comment]
So, this issue is resolved by a PR on another issue from @lazar-stamenkovic? We don't need to do this https://github.com/Expensify/App/issues/21603#issuecomment-1634670845 because we've already done it?

I think this would sadly be a duplicate, rather than qualify for double payment.

If you still need to push a PR, I agree we can assign you and pay you for that. But if one of your PR's happens to have solved two issues, we'll still only pay for one block of work.

**Note** sometimes if the work was large or difficult or complex etc, we do increase the bounty on that, but remember this is always at our discretion. I will certainly take yours and @allroundexperts' opinion into account though if you sincerely believe this is worth more.

**Comment by lazar-stamenkovic** [comment]
@twisterdotcom @allroundexperts Thanks a lot for the accurate clarifications.

It seems that the bug I found is related to the issue described in this ticket. Both problems have a common underlying cause.
The bug reported on this ticket was luckily resolved during a code migration, although the migration was not specifically aimed at fixing this ticket.
I believe that the PR for my proposed solution will address both of these issues effectively.

**Comment by twisterdotcom** [comment]
Okay. I kind of think if we still need to fix something, even if it's not _exactly_ what the OP says, but related, let's just fix it here, related to this issue, and pay out here for 1 issue.

That would still just be 1 issue and 1 payment though.

**Comment by lazar-stamenkovic** [comment]
Thanks! @twisterdotcom @allroundexperts

**Comment by lazar-stamenkovic** [comment]
@deetergp would you please review my PR #22645 ?

**Comment by lazar-stamenkovic** [comment]
@twisterdotcom @allroundexperts @deetergp The PR #22645 has been successfully merged and deployed for this ticket. Could you please provide me with information on how the payment process will proceed? 
Please let me know if there are any updates required on my end.

I would appreciate it if you could consider a bonus for this ticket. 
I managed to create the pull request in 2 days and promptly addressed all the comments received. 
Additionally, I discovered and resolved a new bug during the process for this ticket.

**Comment by twisterdotcom** [comment]
Summary of payments:
- @lazar-stamenkovic has been paid via Upwork $1000: https://www.upwork.com/nx/wm/workroom/34160414/overview
- @allroundexperts is due $1000 via newDot
- @tranvantoan-qn has been sent an offer for $250 here: https://www.upwork.com/nx/wm/offer/25884344

**Comment by JmillsExpensify** [comment]
Reviewed details for @allroundexperts. These details are accurate based on summary from Business Reviewer and are now approved for payment in NewDot.

**Comment by twisterdotcom** [comment]
@allroundexperts could we please have the checklist completed? We can't close the issue until those steps are done.

**Comment by deetergp** [comment]
I don't think we are waiting on anything at this point are we?

**Comment by twisterdotcom** [comment]
Sorry just this: https://github.com/Expensify/Expensify/issues/306327
